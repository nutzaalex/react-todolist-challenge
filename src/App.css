.container {
    max-width: 800px;
    margin: 0 auto;
    font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
}

.error {
    background-color: #ba0000;
    text-align: center;
    color: white;
    padding: 10px;
}