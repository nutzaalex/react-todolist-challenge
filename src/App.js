import React, {Component} from 'react';
import {connect} from 'react-redux';

import List from './components/list';
import Input from './components/input';
import Button from './components/button';

import {fetchInitialList, addItemToList, deleteItemFromList, updateItemInList} from './actions/listActions';

import './App.css';

class App extends Component {
    constructor(props) {
        super(props);

        this.state = {newTodo: ''};

        this.handleChange = this.handleChange.bind(this);
        this.onClick = this.onClick.bind(this);
    }

    componentDidMount() {
        this.props.fetchInitialList();
    }

    handleChange(value) {
        this.setState(() => ({
            newTodo: value,
        }));
    }

    onClick(e) {
        e.preventDefault();
        if (this.state.newTodo) {
            this.props.addItemToList(this.state.newTodo);
            this.setState(() => ({
                newTodo: '',
            }));
        }
    }

    render() {
        const {error, todoList, deleteItemFromList, updateItemInList} = this.props;
        return (
            <div className="container">
                {error ? (<div className="error">{error}</div>) : null}
                <h1>Todo List</h1>

                <div className='add-item-to-list'>
                    <Input
                        name='item'
                        placeholder='New Item...'
                        value={this.state.newTodo}
                        onChange={this.handleChange}
                    />
                    <Button onClick={this.onClick} type='add'>
                        Add
                    </Button>
                </div>

                <List
                    todoList={todoList}
                    deleteItem={deleteItemFromList}
                    updateItem={updateItemInList}
                />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    todoList: state.list.todoList,
    error: state.list.error,
});

const mapDispatchToProps = {fetchInitialList, addItemToList, deleteItemFromList, updateItemInList};

export default connect(mapStateToProps, mapDispatchToProps)(App);

