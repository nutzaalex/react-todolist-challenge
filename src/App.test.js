import React from 'react';
import {mount, shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import App from './App';

const todoList = [{id: 0, item: 'test1'}, {id: 1, item: 'test2'}, {id: 2, item: 'test3'}];

jest.mock('react-redux', () => ({
    connect: () => HOC => HOC,
}));

describe('App', () => {
    it('renders with default props', () => {
        const comp = renderer.create(<App todoList={todoList} fetchInitialList={jest.fn()}/>);

        expect(comp.toJSON()).toMatchSnapshot();
    });
    it('renders with max props', () => {
        const comp = renderer.create(<App todoList={todoList}
                                          error={'some error text'}
                                          fetchInitialList={jest.fn()}
                                          addItemToList={jest.fn()}
                                          deleteItemFromList={jest.fn()}
                                          updateItemInList={jest.fn()}
        />);

        expect(comp.toJSON()).toMatchSnapshot();
    });
    it('fetches the todo list', () => {
        const fetchInitialListSpy = jest.fn();
        shallow(<App todoList={todoList} fetchInitialList={fetchInitialListSpy}/>);

        expect(fetchInitialListSpy).toHaveBeenCalledTimes(1);
    });
    it('calls addItemToList when user clicks on the add button', () => {
        const addItemToListSpy = jest.fn();
        const comp = mount(
            <App
                todoList={todoList}
                fetchInitialList={jest.fn()}
                addItemToList={addItemToListSpy}
            />
        );
        const newTodo = 'new todo';
        comp.find('input').simulate('change', {target: {value: newTodo}});

        expect(comp.find('input').prop('value')).toBe(newTodo);
        expect(addItemToListSpy).not.toHaveBeenCalled();

        comp.find('[type="add"]').simulate('click');

        expect(addItemToListSpy).toHaveBeenCalledTimes(1);
        expect(addItemToListSpy).toHaveBeenCalledWith(newTodo);
    });
    it("updates the input text while typing", () => {
        const comp = mount(<App todList={todoList} fetchInitialList={jest.fn()}/>);
        const newTodo = 'new todo';
        comp.find('input').simulate('change', {target: {value: newTodo}});

        expect(comp.find('input').prop('value')).toBe(newTodo);
    });
});