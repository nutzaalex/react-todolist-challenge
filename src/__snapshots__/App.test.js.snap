// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders with default props 1`] = `
<div
  className="container"
>
  <h1>
    Todo List
  </h1>
  <div
    className="add-item-to-list"
  >
    <input
      className={undefined}
      name="item"
      onChange={[Function]}
      placeholder="New Item..."
      type="text"
      value=""
    />
    <button
      className="button add"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <table>
    <thead>
      <tr>
        <th
          width="66%"
        >
          Item
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          test1
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          test2
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          test3
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`App renders with max props 1`] = `
<div
  className="container"
>
  <div
    className="error"
  >
    some error text
  </div>
  <h1>
    Todo List
  </h1>
  <div
    className="add-item-to-list"
  >
    <input
      className={undefined}
      name="item"
      onChange={[Function]}
      placeholder="New Item..."
      type="text"
      value=""
    />
    <button
      className="button add"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <table>
    <thead>
      <tr>
        <th
          width="66%"
        >
          Item
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          test1
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          test2
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          test3
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
